Multi tenancy support

simon-mo opened the pull request:
This is a work in process, almost done.

Resolve issue #460

dcrankshaw committed:
started impl, but need to dynamically set labels for prom and svcs

simon-mo committed:
Added unbound port func, still debugging

simon-mo committed:
Remove all occurance of tmp in metrics

Resolve issue #448

simon-mo committed:
Resolve hard coded prometheus path

Should resolve issue #424

simon-mo committed:
Add label selection to replicas

simon-mo committed:
Docker Multi-tenancy done

simon-mo committed:
Add templating to config yamls; add cluster_name

simon-mo committed:
Seperation of service done; need check conn

simon-mo committed:
Format Code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1518/
Test FAILed.

simon-mo committed:
Add tests, fix label selector

simon-mo committed:
Format code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1519/
Test FAILed.

simon-mo commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1520/
Test FAILed.

simon-mo committed:
Don't connect after cleanup

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1521/
Test FAILed.

simon-mo committed:
Format code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1522/
Test FAILed.

dcrankshaw commented on bin/run_unittests.sh:
```diff
@@ -144,6 +144,7 @@ function run_integration_tests {
   ../integration-tests/r_integration_test/rclipper_test.sh
   python ../integration-tests/clipper_metric_docker.py 
   python ../integration-tests/clipper_metric_kube.py 
+  python ../integration-tests/multi_tenency_test.py
```
Nit: tenancy is spelled with an "a"

dcrankshaw commented on clipper_admin/clipper_admin/clipper_admin.py:
```diff
@@ -30,7 +30,7 @@
 DEFAULT_LABEL = []
 DEFAULT_PREDICTION_CACHE_SIZE_BYTES = 33554432
-CLIPPER_TEMP_DIR = "/tmp/clipper"
+CLIPPER_TEMP_DIR = "/tmp/clipper"  # Used Internally for Test; Not Windows Compatible
```
Is this true that it's only used for testing?

simon-mo replied:
There's one more occurrence of /tmp/clipper in deployer_utils (just fixed it). It is used to save python function so it's on our critical path. The python code is now fully windows compatible; however the R code still ned to fix. 

```
./containers/R/rclipper_user/R/build.R:  base_model_path = "/tmp/clipper/"
```

dcrankshaw commented on clipper_admin/clipper_admin/container_manager.py:
```diff
@@ -1,19 +1,32 @@
+CLIPPER_INTERNAL_REDIS_PORT = 6379
 
 CLIPPER_DOCKER_LABEL = "ai.clipper.container.label"
+CLIPPER_NAME_LABEL = "ai.clipper.name"
```
This refers to the cluster name? Can we change it to `ai.clipper.cluster.name`?

simon-mo replied:
No this refers to `ai.clipper.name` label in all our kubernetes yaml file. Cluster name set using `CLIPPER_DOCKER_LABEL` as key. (I guess we should change it in the next round of refactoring?)

dcrankshaw commented on clipper_admin/clipper_admin/container_manager.py:
```diff
@@ -1,19 +1,32 @@
 
+CLIPPER_DOCKER_PORT_LABELS = {
+    'redis': 'ai.clipper.redis.port',
+    'query_query': 'ai.clipper.query_frontend.query.port',
```
What is query_query? Can you use a more descriptive name here?

dcrankshaw replied:
It looks like this is for the client-facing service. Can you call it `query_rest`?

simon-mo replied:
`query_rest` is a great name!

dcrankshaw commented on clipper_admin/clipper_admin/container_manager.py:
```diff
@@ -34,6 +47,50 @@ def parse_model_container_label(label):
 
+def find_unbound_port(start=None,
+                      increment=True,
+                      port_range=(34256, 50000),
```
Why did you pick 34256 as the lower bound on the port range?

simon-mo replied:
idk no special reason it comes from previous code; i'll do 10000

dcrankshaw commented on clipper_admin/clipper_admin/container_manager.py:
```diff
@@ -34,6 +47,50 @@ def parse_model_container_label(label):
     return splits
 
 
+def find_unbound_port(start=None,
```
Move this function into the docker container manager, as it only works for finding locally unbound ports. The K8s container shouldn't use it (or need to).

dcrankshaw commented on clipper_admin/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -115,22 +125,43 @@ def start_clipper(self,
+        containers_in_cluster = self.docker_client.containers.list(
+            filters={
+                'label':
+                ['ai.clipper.container.label={}'.format(self.cluster_name)]
```
Don't use the literal string here for the label key, use a variable instead

dcrankshaw replied:
Reminder to address this comment

dcrankshaw commented on clipper_admin/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -115,22 +125,43 @@ def start_clipper(self,
+        if len(containers_in_cluster) > 0:
+            raise ClipperException(
+                "Cluster {} cannot be started because it already exists. "
+                "Please use clipper_conn.connect() to connect to it.".format(
```
s/clipper_conn.connect()/ClipperConnection.connect()

dcrankshaw commented on clipper_admin/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -115,22 +125,43 @@ def start_clipper(self,
+            self.redis_port = find_unbound_port(self.redis_port)
+            redis_labels = self.common_labels.copy()
+            redis_labels[CLIPPER_DOCKER_PORT_LABELS['redis']] = str(
+                self.redis_port)
```
What is this port label used for?

simon-mo replied:
TL;DR: useful for ClipperConnection.connect()

Because the port is now floating, we no longer have knowledge of port mapping. So when user calls connect, we will know which port correspond to which container.

There will an alternative by just using docker python client but requires a "three way join" among docker containers, docker network, and some docker client magics; current approach is more straightforward to implement and understand.

simon-mo committed:
Add debug line; tests pass locally; suspecting dangling containers

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1524/
Test FAILed.

dcrankshaw commented on clipper_admin/clipper_admin/container_manager.py:
```diff
@@ -34,6 +47,50 @@ def parse_model_container_label(label):
+                      verbose=False,
+                      logger=None):
+    """
+    Fina a unbound port.
```
Nit: s/Fina/Find

simon-mo committed:
Checkout parallel build docker

simon-mo committed:
Make persistent test less verbose

simon-mo committed:
Make Unittest all parallel

simon-mo committed:
Make maven less verbose

simon-mo committed:
Change persistent_state_test 10->100

simon-mo committed:
Refactor tests to include cluster name

simon-mo committed:
Format code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1533/
Test FAILed.

simon-mo committed:
Add logging support; push to test on ubuntu

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1537/
Test FAILed.

simon-mo committed:
Fix docker issue on ubuntu

simon-mo committed:
Modify comment about prometheus

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1538/
Test FAILed.

simon-mo committed:
Fix some kubernetes test

simon-mo committed:
Interpreable cluster name

simon-mo committed:
Format code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1539/
Test FAILed.

simon-mo commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1543/
Test FAILed.

dcrankshaw commented on bin/run_unittests.sh:
```diff
@@ -91,7 +91,7 @@ function run_jvm_container_tests {
   cd $DIR
   cd ../containers/jvm
-  mvn test
+  mvn test -q
```
👍

dcrankshaw commented on clipper_admin/clipper_admin/clipper_admin.py:
```diff
@@ -447,17 +453,18 @@ def build_model(self,
                 fileobj=context_file, custom_context=True, tag=image)
             for b in build_logs:
-                logger.info(b)
+                if 'stream' in b and b['stream'] != '\n':  #log build steps only
```
👍

dcrankshaw commented on clipper_admin/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -146,12 +181,19 @@ def start_clipper(self,
+        self.clipper_query_port = find_unbound_port(self.clipper_query_port)
+        self.clipper_rpc_port = find_unbound_port(self.clipper_rpc_port)
+        query_labels = self.common_labels.copy()
+        query_labels[CLIPPER_QUERY_FRONTEND_CONTAINER_LABEL] = ""
```
Why is this label value empty?

simon-mo replied:
It's used to pin down specific container:

https://github.com/ucbrise/clipper/blob/1394de0867ac53efa5ed601bd91806d93d2f3ef4/clipper_admin/clipper_admin/docker/docker_container_manager.py#L211

dcrankshaw replied:
Hmm I'm still not sure I'm following. How exactly is this label used?

simon-mo replied:
It was used here: https://github.com/ucbrise/clipper/blob/062b968f96c3821a354374f0b90d9c3776618419/clipper_admin/clipper_admin/docker/docker_container_manager.py#L207-L218

to search for live query frontend containers. The value is empty simply because we don't need to add anything. "ai.clipper.query_frontend.label" is enough.

This was introduced from last summer's big refactoring: https://github.com/ucbrise/clipper/pull/251

dcrankshaw commented on clipper_admin/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -170,16 +212,49 @@ def start_clipper(self,
+        containers = self.docker_client.containers.list(
+            filters={
+                'label':
+                ['ai.clipper.container.label={}'.format(self.cluster_name)]
```
Don't use the literal value of the label, use a variable to reference it instead

dcrankshaw replied:
Reminder

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/kubernetes_container_manager.py:
```diff
@@ -62,6 +91,9 @@ def __init__(self,
             if no ExternalDNS exists for any of the nodes.
+        kubernetes_namespace : str, optional
+            Which Kubernetes namespace to start the Clipper cluster in. The namespace must already exist.
+            If no namespace is provided, Clipper will be started in the default namespace.
```
Note: Make sure to rebase this on #500

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/kubernetes_container_manager.py:
```diff
@@ -85,103 +119,133 @@ def __init__(self,
+
+    def _start_mgmt(self, mgmt_image):
+        with _pass_conflicts():
+            mgmt_depolyment_data = self._generate_config(
```
nit: mispelled deployment here

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/kubernetes_container_manager.py:
```diff
@@ -85,103 +119,133 @@ def __init__(self,
+                redis_service_port=self.redis_port,
+                cluster_name=self.cluster_name)
+            self._k8s_beta.create_namespaced_deployment(
+                body=mgmt_depolyment_data, namespace='default')
```
Missed a 'default' namespace reference. It's definitely worth grepping for "default" in this file to double check you didn't miss any (although I know #500 takes care of this for us).

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/kubernetes_container_manager.py:
```diff
@@ -85,103 +119,133 @@ def __init__(self,
+            self._k8s_v1.create_namespaced_service(
+                body=service_data, namespace=self.namespace)
+
+    def _generate_config(self, file_path, **kwargs):
```
Nice idea on using a template engine. This is much cleaner.

simon-mo replied:
This is an intermediate step. In the future we should adopt helm chart.

dcrankshaw replied:
Helm is definitely worth looking into. Our Kubernetes cluster is a bit more dynamic than most (as we dynamically create and delete deployments and services). The right way to do it is probably to use a Helm chart to start the initial cluster, then not use Helm for creating model deployments. This is probably worth looking into when we move to having the active API server managing containers for us.

The other thing to keep in mind is that we want to be compatible with different container orchestration systems, so we never want to only support Kubernetes.

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/kubernetes_container_manager.py:
```diff
@@ -250,85 +326,30 @@ def connect(self):
+            with _pass_conflicts():
                 self._k8s_beta.create_namespaced_deployment(
-                    body=body, namespace='default')
+                    body=generated_body, namespace='default')
```
Missed a reference to default namespace

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/kubernetes_container_manager.py:
```diff
@@ -250,85 +326,30 @@ def connect(self):
-                        != num_replicas:
-                    time.sleep(3)
+            while self._k8s_beta.read_namespaced_deployment_status(
+                name=deployment_name, namespace='default').status.available_replicas \
```
Missed a reference to default namespace

dcrankshaw commented on integration-tests/multi_tenency_test.py:
```diff
@@ -0,0 +1,78 @@
+from clipper_admin import ClipperConnection, DockerContainerManager, KubernetesContainerManager
```
Nit: mispelled multi-tenancy (s/tenency/tenancy)

dcrankshaw reviewed (changes requested):
This looks great. Just had some minor comments.

dcrankshaw commented:
@simon-mo can you add a test for running Clipper in a user-specified namespace as well?

dcrankshaw commented:
Another place where we will run into contention issues is the remote docker registry for the kubernetes tests. We should tag the images we push with the commit hash, rather than using `develop`.

simon-mo committed:
Merge remote-tracking branch 'upstream/develop' into multi_tenancy_support

simon-mo committed:
Remove temp_dir usage in deployer serialize func

simon-mo commented:
@dcrankshaw 

- test added
- Contention issue goes much deeper with multi-tenancy & namespace. When user deploy models with the same name & version tag under different cluster name, we are tagging them the same way. However, I don't think this is something to address in this PR. We can say multi-tenancy support is experimental for now and fix this type of issue in management re-write.

simon-mo committed:
Address comment

simon-mo committed:
Format code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1568/
Test FAILed.

simon-mo committed:
Add more comments; shorten naming for cluster

simon-mo committed:
Merge branch 'develop' into multi_tenancy_support

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1571/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1572/
Test FAILed.

dcrankshaw commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1574/
Test FAILed.

simon-mo committed:
Merge branch 'develop' into multi_tenancy_support

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1575/
Test FAILed.

simon-mo commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1576/
Test FAILed.

simon-mo committed:
Fix proxy addr

simon-mo committed:
Format code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1577/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1578/
Test FAILed.

simon-mo committed:
Wrap k8s metric test 2 in retry loop

This is the only way to prevent some jenkins & mismatch

simon-mo committed:
Clean up test

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1580/
Test FAILed.

simon-mo committed:
Merge branch 'develop' of https://github.com/ucbrise/clipper into multi_tenancy_support

simon-mo committed:
Format code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1581/
Test FAILed.

simon-mo commented:
There are few action items to move this PR forward:
- [x]  Change version tag in the unittest container (in `build_docker.sh`) from `develop` to git commit hash
- [x] In model deployment, image building and tagging should include the cluster-name to prevent name collision within the same registry
- [x] Change k8s docker registry from AWS to dockerhub using clipper's test account

simon-mo committed:
Fix frontend exporter naming

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1585/
Test FAILed.

simon-mo committed:
Add readme to CI process

simon-mo committed:
Change testapp name for ekr

simon-mo committed:
Swap the python closure build step

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1589/
Test PASSed.

simon-mo committed:
Add __registry__ argument

simon-mo committed:
Add registry to docker build process

simon-mo committed:
Add cluster info to image name

simon-mo committed:
Merge branch 'develop' of https://github.com/ucbrise/clipper into multi_tenancy_support

simon-mo committed:
Format code

simon-mo commented:
@dcrankshaw This PR is ready for another round of review now.

simon-mo committed:
Run unittests in clippertest

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1590/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1591/
Test FAILed.

dcrankshaw commented on bin/README.md:
```diff
@@ -0,0 +1,21 @@
+## How does the CI process work
+
+
+0. Jenkins pull the PR and sandbox it.
```
Jenkins actually doesn't do any sandboxing.

dcrankshaw commented on bin/README.md:
```diff
@@ -0,0 +1,21 @@
+
+
+0. Jenkins pull the PR and sandbox it.
+1. Jenkins inject environment variables configured in admin page
```
Let's list what environment variables are set

dcrankshaw commented on bin/build_docker_images.sh:
```diff
@@ -211,9 +211,14 @@ create_image () {
-    docker tag $namespace/$image:$sha_tag $namespace/$image:$version_tag
+
+    echo "Publishing $namespace/$image:$sha_tag from file $dockerfile"
+    docker push $namespace/$image:$sha_tag
```
There shouldn't be a docker push here. We only want to push images if `publish` is True.

dcrankshaw commented on bin/build_docker_images.sh:
```diff
@@ -211,9 +211,14 @@ create_image () {
+    docker push $namespace/$image:$sha_tag
+
+    # We will NOT tag the image to version tag to prevent collision
+    # docker tag $namespace/$image:$sha_tag $namespace/$image:$version_tag
```
You still need this tag call to publish, just move it inside the `if publish` block.

dcrankshaw commented on bin/run_ci.sh:
```diff
@@ -21,12 +21,7 @@ tag=$(<VERSION.txt)
-docker tag clipper/query_frontend:$tag 568959175238.dkr.ecr.us-west-1.amazonaws.com/clipper/query_frontend:$tag
-docker push 568959175238.dkr.ecr.us-west-1.amazonaws.com/clipper/query_frontend:$tag
-docker tag clipper/management_frontend:$tag 568959175238.dkr.ecr.us-west-1.amazonaws.com/clipper/management_frontend:$tag
-docker push 568959175238.dkr.ecr.us-west-1.amazonaws.com/clipper/management_frontend:$tag
```
We don't want the docker build images script to automatically push the images. Remember, that script is very useful during development for rebuilding images locally. Instead, let's change `build_docker_images.sh` to write the names of all the containers it builds to a file (something like `clipper_docker_images.txt`) and then you can push all those images with the following one-liner (from https://stackoverflow.com/a/13939059/814642):

```
while read in; do docker push "$in"; done < clipper_docker_images.txt
```

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/kubernetes_container_manager.py:
```diff
@@ -146,20 +156,29 @@ def __init__(self,
 
+        self.cluster_identifier = "{ns}{cluster}".format(
+            ns=self.k8s_namespace + '-'
+            if self.k8s_namespace != "default" else "",
```
This syntax is pretty convoluted. Can you write this as just a normal if-statement outside the format argument (for maintainability reasons)?

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/kubernetes_container_manager.py:
```diff
@@ -277,46 +297,46 @@ def connect(self):
         if len(external_node_hosts) == 0:
-            msg = "Error connecting to Kubernetes cluster. No external node addresses found"
-            logger.error(msg)
+            msg = "Error connecting to Kubernetes cluster. No external node addresses found. If you are running Kubernetes locally, you can pass in KubernetesContainerManager(useInternalIP=True) to connect to local Kubernetes cluster"
```
You don't need to be running Kubernetes locally to use internal ip addresses. Remove that part of the error message.

dcrankshaw commented on clipper_admin/clipper_admin/kubernetes/kubernetes_container_manager.py:
```diff
@@ -414,8 +435,8 @@ def get_logs(self, logging_dir):
-                logger.info("log file name: {}".format(log_file_name))
-                logger.info("log alt file name: {}".format(log_file_alt))
+                self.logger.info("log file name: {}".format(log_file_name))
+                self.logger.info("log alt file name: {}".format(log_file_alt))
```
What is the purpose of `log_file_alt`? It doesn't look like it's being used at all.

simon-mo replied:
frankly i have no idea. I just deleted this line

dcrankshaw commented on dockerfiles/ClipperLibBaseDockerfile:
```diff
@@ -1,3 +1,4 @@
+ARG REGISTRY
```
Move this below the comment, as the `REGISTRY` arg isn't being used either

dcrankshaw commented on dockerfiles/FrontendExporterDockerfile:
```diff
@@ -1,3 +1,4 @@
+ARG REGISTRY
```
Move below unused arg comment

dcrankshaw commented on dockerfiles/Py2RPCDockerfile:
```diff
@@ -1,3 +1,4 @@
+ARG REGISTRY
```
Move below unused arg comment

dcrankshaw commented on dockerfiles/Py35RPCDockerfile:
```diff
@@ -1,3 +1,4 @@
+ARG REGISTRY
```
Move below unused arg comment

dcrankshaw commented on dockerfiles/Py36RPCDockerfile:
```diff
@@ -1,3 +1,4 @@
+ARG REGISTRY
```
Move below unused arg comment

dcrankshaw commented on dockerfiles/RContainerDockerfile:
```diff
@@ -1,3 +1,4 @@
+ARG REGISTRY
```
Move below unused arg comment

dcrankshaw commented on dockerfiles/SparkScalaContainerDockerfile:
```diff
@@ -1,3 +1,4 @@
+ARG REGISTRY
```
Move below unused arg comment

dcrankshaw reviewed (changes requested):
Nice job. Almost there.

simon-mo committed:
Address comment

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1592/
Test FAILed.

simon-mo committed:
fix clipper_docker_images location

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1593/
Test FAILed.

simon-mo commented on clipper_admin/clipper_admin/docker/docker_container_manager.py:
```diff
@@ -169,17 +218,55 @@ def start_clipper(self,
+        containers = self.docker_client.containers.list(
+            filters={
+                'label': [
+                    '{key}={val}'.format(
```
@dcrankshaw the literal label is fixed here (same above). I don't know why github is still showing the old diff

simon-mo committed:
Add set -x to debug "invalid reference format"

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1594/
Test FAILed.

simon-mo committed:
Add debug lines

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1595/
Test FAILed.

simon-mo committed:
Fix the blank line issue in clipper_docker_images.txt

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1596/
Test FAILed.

simon-mo committed:
Don't pull ecr images

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1597/
Test FAILed.

simon-mo committed:
Fix import

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1598/
Test FAILed.

simon-mo committed:
Set git hash length to at least 10 to prevent collision

Such that different version of git in different environement can agree
that the length of the hash

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1599/
Test FAILed.

simon-mo committed:
Update registry in test

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1600/
Test FAILed.

simon-mo committed:
Add version back to metrics

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1602/
Test FAILed.

simon-mo committed:
Remove CLIPPER_CONTAINER_REGISTRY constant in test

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1603/
Test FAILed.

simon-mo committed:
Revert "Remove CLIPPER_CONTAINER_REGISTRY constant in test"

This reverts commit 99e25172b8520872a2dea6f526b083d7c7fda7a3.

simon-mo committed:
Add CLIPPER_CONTAINER_REGISTRY back

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1605/
Test FAILed.

simon-mo committed:
Format code

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1606/
Test FAILed.

simon-mo commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1607/
Test FAILed.

simon-mo committed:
Fix docker login in unnitest

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1608/
Test FAILed.

simon-mo commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1610/
Test FAILed.

simon-mo commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1612/
Test FAILed.

simon-mo committed:
Fix image name in R test

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1613/
Test FAILed.

simon-mo commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1614/
Test FAILed.

simon-mo committed:
Fix CLIPPER_CONTAINER_REGISTRY in k8s metrics test

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1615/
Test FAILed.

simon-mo commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1616/
Test FAILed.

simon-mo commented:
jenkins test this please

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1617/
Test FAILed.

simon-mo committed:
Refresh k8s metric tests;fix variable reuse

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Clipper-PRB/1618/
Test PASSed.

dcrankshaw reviewed (approved):
LGTM
